### PR TITLE
Fix #3965, before/after upsert hooks

### LINKED ISSUE #3965: before/after upsert hook
I'm using PostgreSQL and noticed that the Update/Create hooks are not being triggered by model upserts.

```javascript
// Example
    var User = sequelize.define('User', {
        name: sequelize.Sequelize.STRING
    });

   User.hook('beforeCreate', function (user, options) {
        console.log('In beforeCreate hook.');
        return Promise.resolve(user);
    });

   User.hook('beforeUpdate', function (user, options) {
        console.log('In beforeUpdate hook.');
        return Promise.resolve(user);
    });

   User.hook('afterCreate', function (user, options) {
        console.log('In afterCreate hook.');
        return Promise.resolve(user);
    });

   User.hook('afterUpdate', function (user, options) {
        console.log('In afterUpdate hook.');
        return Promise.resolve(user);
    });

    User
        .upsert({name: 'Pedro'}) // Insert
        .then(function (created) {
            console.log(created); // true
            return User.upsert({id: 1, name: 'Pablo'}); // Update
        })
        .then(function (created) {
            console.log(created); // false
        });
```
None of the hooks are called. Is there a way around this? This is under sequelize v2.1.3.

SQL for the upsert:
```
// Insert upsert.
Executing (default): CREATE OR REPLACE FUNCTION pg_temp.sequelize_upsert() RETURNS integer AS $$ BEGIN INSERT INTO "Users" ("createdAt","updatedAt") VALUES ('2015-06-18 19:14:13.247 +00:00','2015-06-18 19:14:13.247 +00:00'); RETURN 1; EXCEPTION WHEN unique_violation THEN UPDATE "Users" SET "updatedAt"='2015-06-18 19:14:13.247 +00:00'; RETURN 2; END; $$ LANGUAGE plpgsql; SELECT * FROM pg_temp.sequelize_upsert();

// Update upsert.
Executing (default): CREATE OR REPLACE FUNCTION pg_temp.sequelize_upsert() RETURNS integer AS $$ BEGIN INSERT INTO "Users" ("id","createdAt","updatedAt") VALUES (1,'2015-06-18 19:14:13.254 +00:00','2015-06-18 19:14:13.254 +00:00'); RETURN 1; EXCEPTION WHEN unique_violation THEN UPDATE "Users" SET "id"=1,"updatedAt"='2015-06-18 19:14:13.254 +00:00' WHERE ("id" = 1); RETURN 2; END; $$ LANGUAGE plpgsql; SELECT * FROM pg_temp.sequelize_upsert();
```

### PR DESCRIPTION
**Pull request by sushantdhiman**
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

### Description of change

Closes #3965 

Added `before/after Upsert` hooks with tests and documentation update.

### CODE REVIEW

File: lib/model.js [hunk 1]
```diff
@@ -2006,7 +2009,17 @@ class Model {
         delete updateValues[this.primaryKeyField];
       }
 
-      return this.QueryInterface.upsert(this.getTableName(options), insertValues, updateValues, instance.where(), this, options);
+      return new Promise((resolve) => {
+        if (options.hooks) {
+          resolve(this.runHooks('beforeUpsert', values, options));
+        } else { resolve(); }
+      }).then(() => (
```
Comment (mickhansen) [inline lib/model.js]: `Promise.try` is a bit nicer than `new Promise`

Comment (sushantdhiman) [inline lib/model.js]: Great, Learned something new

**Comment by sushantdhiman** [comment]
@mickhansen anything else ?

**Comment by jdnichollsc** [comment]
Hi guys! What is the version of Sequelize that include this fix? 👍

**Comment by sushantdhiman** [comment]
`v4.0.0-1`

**Comment by alexdrans** [comment]
Publish to npm?

**Comment by mickhansen** [comment]
@alexdrans I believe it is, what makes you think it's not?

**Comment by alexdrans** [comment]
``3.24.3`` is the latest on npm from what I can see?

**Comment by janmeier** [comment]
Latest stable yes - This is publish in v4, which is not yet stable. You can install v4 using `npm i sequelize@4` or `npm i sequelize@unstable`

**Pull request merged**